### PR TITLE
`not` combinator

### PR DESCRIPTION
**Pull request by bencochran**
Implements a negative lookahead combinator as described in #51.

I started with it called [`notFollowedBy`](https://hackage.haskell.org/package/parsec-3.1.7/docs/src/Text-Parsec-Combinator.html#notFollowedBy), but most of my uses look something like `many(not(something) *> any)` and having `notFollowedBy(something) *> any` felt confusing.

This could also yield a subtraction combinator (#59) via `not(right) *> left` if we want one.

### CODE REVIEW

**Comment by robrix** [comment]
This is great, thank you!

**Comment by robrix** [comment]
I wonder if we should document that you can run into trouble with e.g.:

```swift
many(not(something))
```

Since `not` doesn’t advance the cursor, it can infinite loop. Alternatively, maybe `many` should not repeat at the same index.

Any thoughts?

**Comment by bencochran** [comment]
That’s a good catch.

Poked around at making `many` (and therefore also `some`) fail if they don’t consume any input. I’m not convinced it’s the cleanest approach, so I may come take a second look at it, but wanted to get it up here for others’ eyes in the meantime.

File: MadnessTests/RepetitionTests.swift [hunk 1]
```diff
@@ -191,8 +191,25 @@ final class RepetitionTests: XCTestCase {
...
+
+	let oneOrMoreLookAhead = some(not(%"x"))
+
+	func testOneOrMoreFailsLookAhead() {
+		assertUnmatched(oneOrMoreLookAhead, "yyy".characters)
+	}
 }
 
+private func == (left: [()], right: [()]) -> Bool {
+	return left.count == right.count
```
Comment (robrix) [inline MadnessTests/RepetitionTests.swift]: :+1:

File: MadnessTests/RepetitionTests.swift [hunk 2]
```diff
@@ -191,8 +191,25 @@ final class RepetitionTests: XCTestCase {
...
+	func testZeroOrMoreSkipsLookAhead() {
+		assertTree(zeroOrMoreLookAhead, "yyy".characters, ==, [])
+	}
+
+
+	let oneOrMoreLookAhead = some(not(%"x"))
+
+	func testOneOrMoreFailsLookAhead() {
+		assertUnmatched(oneOrMoreLookAhead, "yyy".characters)
+	}
```
Comment (robrix) [inline MadnessTests/RepetitionTests.swift]: :sparkles: these tests are excellent, thank you!

Comment (bencochran) [inline MadnessTests/RepetitionTests.swift]: :grinning: :bow:

**Comment by bencochran** [comment]
Ok, coming back for a second ponder: the current approach is sitting alright with me. The only hesitation I have is wondering if an error is the correct choice. That is, I could see an argument that `some(not(x))` should match once then bail (that is, it would effectively produce `[()]`). But that feels like it would be uglier to implement, and I’m not convinced there’s a real reason for that behavior above this one.

**Comment by robrix** [comment]
:+1: Thanks @bencochran. This is looking really good.

**Comment by bencochran** [comment]
Awesome, thank you!

**Pull request merged**